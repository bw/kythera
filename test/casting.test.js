const t = require("./util").test

describe("Type casting", () => {
	describe("Custom type assignment", () => {
		t("primitive", `let myIntType = int`)
		t("fn type", `let myFnType = fn<int, str, myIntType> str`)
	})
})