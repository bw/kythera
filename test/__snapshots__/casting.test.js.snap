// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type casting [PARSE] Casting object 1`] = `
Array [
  ParseNode {
    "identifier": "myObj",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "obj",
        "kind": "type",
        "origin": "builtin",
        "structure": Object {},
      },
      "value": Object {
        "a": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          "value": 99,
        },
        "b": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
          "value": "beep",
        },
      },
    },
  },
  ParseNode {
    "index": "a",
    "kind": "access",
    "method": "dot",
    "target": ParseNode {
      "kind": "identifier",
      "name": "myObj",
    },
  },
  ParseNode {
    "index": "a",
    "kind": "access",
    "method": "dot",
    "target": ParseNode {
      "from": ParseNode {
        "kind": "identifier",
        "name": "myObj",
      },
      "kind": "as",
      "to": ParseNode {
        "kind": "type",
        "name": "myObjType",
        "origin": "named",
      },
    },
  },
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "index": "a",
      "kind": "access",
      "method": "dot",
      "target": ParseNode {
        "kind": "identifier",
        "name": "myObj",
      },
    },
    "operator": "+",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "int",
        "kind": "type",
        "origin": "builtin",
      },
      "value": 10,
    },
  },
  ParseNode {
    "identifier": "myCastedObj",
    "kind": "let",
    "value": ParseNode {
      "from": ParseNode {
        "kind": "identifier",
        "name": "myObj",
      },
      "kind": "as",
      "to": ParseNode {
        "kind": "type",
        "name": "myObjType",
        "origin": "named",
      },
    },
  },
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "identifier",
      "name": "myCastedObj",
    },
  },
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "typeof",
      "target": ParseNode {
        "kind": "identifier",
        "name": "myCastedObj",
      },
    },
    "operator": "==",
    "right": ParseNode {
      "kind": "typeof",
      "target": ParseNode {
        "kind": "identifier",
        "name": "myObj",
      },
    },
  },
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "typeof",
      "target": ParseNode {
        "kind": "identifier",
        "name": "myCastedObj",
      },
    },
    "operator": "==",
    "right": ParseNode {
      "kind": "typeof",
      "target": ParseNode {
        "from": ParseNode {
          "kind": "identifier",
          "name": "myObj",
        },
        "kind": "as",
        "to": ParseNode {
          "kind": "type",
          "name": "myObjType",
          "origin": "named",
        },
      },
    },
  },
]
`;

exports[`Type casting [PARSE] Usage as fn param 1`] = `
Array [
  ParseNode {
    "identifier": "myObjType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "obj",
        "kind": "type",
        "origin": "builtin",
        "structure": Object {
          "a": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          "b": ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
        },
      },
    },
  },
  ParseNode {
    "identifier": "myObj",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "obj",
        "kind": "type",
        "origin": "builtin",
        "structure": Object {},
      },
      "value": Object {
        "a": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          "value": 99,
        },
        "b": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
          "value": "beep",
        },
      },
    },
  },
  ParseNode {
    "identifier": "myCastedObj",
    "kind": "let",
    "value": ParseNode {
      "from": ParseNode {
        "kind": "identifier",
        "name": "myObj",
      },
      "kind": "as",
      "to": ParseNode {
        "kind": "type",
        "name": "myObjType",
        "origin": "named",
      },
    },
  },
  ParseNode {
    "identifier": "myFn",
    "kind": "let",
    "value": ParseNode {
      "body": Array [
        ParseNode {
          "index": "a",
          "kind": "access",
          "method": "dot",
          "target": ParseNode {
            "kind": "identifier",
            "name": "object",
          },
        },
        ParseNode {
          "index": "b",
          "kind": "access",
          "method": "dot",
          "target": ParseNode {
            "kind": "identifier",
            "name": "object",
          },
        },
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "baseType": "int",
              "kind": "type",
              "origin": "builtin",
            },
            "value": 10,
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [
        Object {
          "name": "object",
          "type": ParseNode {
            "kind": "type",
            "name": "myObjType",
            "origin": "named",
          },
        },
      ],
      "returns": ParseNode {
        "baseType": "null",
        "kind": "type",
        "origin": "builtin",
      },
      "type": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "kind": "type",
            "name": "myObjType",
            "origin": "named",
          },
        ],
        "returns": ParseNode {
          "baseType": "null",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
  ParseNode {
    "arguments": Array [
      ParseNode {
        "from": ParseNode {
          "kind": "identifier",
          "name": "myObj",
        },
        "kind": "as",
        "to": ParseNode {
          "kind": "type",
          "name": "myObjType",
          "origin": "named",
        },
      },
    ],
    "kind": "call",
    "target": ParseNode {
      "kind": "identifier",
      "name": "myFn",
    },
  },
  ParseNode {
    "arguments": Array [
      ParseNode {
        "kind": "identifier",
        "name": "myCastedObj",
      },
    ],
    "kind": "call",
    "target": ParseNode {
      "kind": "identifier",
      "name": "myFn",
    },
  },
]
`;

exports[`Type casting Custom type assignment [PARSE] fn type 1`] = `
Array [
  ParseNode {
    "identifier": "myFnType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
          ParseNode {
            "kind": "type",
            "name": "myIntType",
            "origin": "named",
          },
        ],
        "returns": ParseNode {
          "baseType": "str",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
]
`;

exports[`Type casting Custom type assignment [PARSE] obj type 1`] = `
Array [
  ParseNode {
    "identifier": "myObjType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "obj",
        "kind": "type",
        "origin": "builtin",
        "structure": Object {
          "a": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          "b": ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
        },
      },
    },
  },
]
`;

exports[`Type casting Custom type assignment [PARSE] primitive 1`] = `
Array [
  ParseNode {
    "identifier": "myIntType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "int",
        "kind": "type",
        "origin": "builtin",
      },
    },
  },
]
`;
